Src/libCZI/utilities.cpp

```diff
@@ -423,3 +424,158 @@ tString trimImpl(const tString& str, const tString& whitespace)
     LoHiBytePackStrided_C(ptrSrc, sizeSrc, width, height, stride, dest);
 }
 #endif
+
+void RectangleCoverageCalculator::AddRectangle(const libCZI::IntRect& rectangle)
+{
+    if (!rectangle.IsValid())
+    {
+        return;
+    }
+
+    if (this->splitters_.empty())
+    {
+        this->splitters_.push_back(rectangle);
+    }
+    else
+    {
+        for (auto i = this->splitters_.begin(); i != this->splitters_.end(); ++i)
+        {
+            // does the rectangle intersect with one of existing rectangles?
+            if (i->IntersectsWith(rectangle) == true)
+            {
+                // check if it is completely contained in the current existing rectangle
+                if (RectangleCoverageCalculator::IsCompletelyContained(*i, rectangle) == true)
+                {
+                    // ok, in this case we have nothing to do!
+                    return;
+                }
+                else
+                {
+                    // check if the existing rectangle is completely contained in the new one
+                    if (RectangleCoverageCalculator::IsCompletelyContained(rectangle, *i) == true)
+                    {
+                        // in this case we remove the (smaller) rect splitters[i] (which is fully
+                        // contained in rectangle) from our list and add rectangle
+                        this->splitters_.erase(i);
+                        this->AddRectangle(rectangle);
+                        return;
+                    }
+
+                    // ok, the rectangle overlap only partially... then let's cut the new rectangle into pieces
+                    // (which do not intersect with the currently investigated rectangle) and try
+                    // to add those pieces
+                    std::array<libCZI::IntRect, 4> splitUpRects;
+                    const int number_of_split_up_rects = SplitUpIntoNonOverlapping(*i, rectangle, splitUpRects);
+                    for (int n = 0; n < number_of_split_up_rects; ++n)
+                    {
+                        this->AddRectangle(splitUpRects[n]);
+                    }
+
+                    return;
+                }
+            }
+        }
+
+        // if we end up here this means that the new rectangle does not
+        // overlap with an existing one -> we can happily add it now!
+        this->splitters_.push_back(rectangle);
+    }
+}
+
+/*static*/int RectangleCoverageCalculator::SplitUpIntoNonOverlapping(const libCZI::IntRect& rectangle_a, const libCZI::IntRect& rectangle_b, std::array<libCZI::IntRect, 4>& result)
```

Comment:
i won't review this, this is logic that should be thoroughly unit-tested (and it seems that it is).

Reply:
Yes, it comes with extensive tests (with a random set of rectangles), and I ran them for a long time successfully.